### ISSUE TITLE
Productize rapidpro2pg

### ISSUE THREAD

**Issue by michaelkohn** [medic/cht-core#6641]
App Services have written a program similar to couch2pg that pulls data from RapidPro and makes it available in PostgreSQL for ease of analysis. We should consider productizing this solution.

**Comment by MaxDiz** [comment]
Following on a [slack string](https://medic.slack.com/archives/C01EESE66MT/p1605134800021900)...Rapidpro was recently deployed in product via medic/cht-core#6379 . For the CIHA deployment, the data push from textit into postgres on the partners side is currently a manual job. This is causing challenges for supporting first class data use.

@kitsao is working on a solution which is in the process of being "dockerized". Do you need additional support from the product team to make this production ready?

**Comment by kitsao** [comment]
Done on the [rapidpro2pg](https://github.com/medic/rapidpro2pg) repo @MaxDiz @michaelkohn

**Comment by MaxDiz** [comment]
sweet! Thanks @kitsao 

@craig-landry do you want to work with @ecsalomon to go through QA as [mentioned here](https://medic.slack.com/archives/C01EESE66MT/p1605134800021900)?

**Comment by MaxDiz** [comment]
Hi @ecsalomon would you be able to do a data quality check to identify any gaps in expectations for this work? 

Eng won't be able to pick this up until v3.12 to do a code review.

**Comment by not-the-fish** [comment]
I will work with @marialma on that.

**Comment by MaxDiz** [comment]
This issue requires 2 items:
1) Data science review of data quality output
2) Engineering review of code quality and scalability. If changes are made by the engineering team, it may be a good idea for Data Science to re-review

**Comment by mrjones-plip** [comment]
@marialma - as we're looking to prepare working on the 3.12 release, are you still the right person to contact for a "Data science review of data quality output"?  If not, could you recommend who would be a good contact instead?

@michaelkohn  - when we say we want to "Productize rapidpro2pg" what does that mean? Does that mean stand up an instance as a peer of couch2pg that writes to our big rdbms instance or instead ship a functional version with every cht-core deployment?  Or something else all together different ;) !

**Comment by marialma** [comment]
Oh, I guess I am.

**Comment by Hareet** [comment]
>  Does that mean stand up an instance ... or instead ship a functional version with every cht-core deployment?

@mrjones-plip 
container container container! 

A separate container handles our current architecture as well as future architectures (v3!)

**Comment by garethbowen** [comment]
In addition to what Hareet said, we probably won't "ship a functional version with every cht-core deployment" in the near term as there aren't that many projects using our RapidPro hosting yet. If it's done as a container then each project can decide if and how they want to deploy it.

**Comment by mrjones-plip** [comment]
> Oh, I guess I am.

@marialma - Awesome, welcome! Can you review the comments/links above to see if you have what you need for a data science review of data quality output?

> Gareth: we probably won't "ship a functional version with every cht-core deployment" in the near term
> Hareet: A separate container handles our current architecture as well as future architectures (v3!)

@Hareet  and @garethbowen  - awesome thanks for the input!  The two end results of this ticket might be a published container (a peer of `medic-os`) with some updates to the docs site on how to deploy/integrate it?  I know we have a couch2pg hiding in current `medic-os`, so that's what made me think "ship a functional version with every cht-core" ;)

**Comment by michaelkohn** [comment]
> when we say we want to "Productize rapidpro2pg" what does that mean

Also means making sure that it includes, or is easy to include, all the endpoints we normally need.  For example, the [current version](https://github.com/medic/rapidpro2pg) only gets data from [3 RapidPro API endpoints](https://github.com/medic/rapidpro2pg#references).  We should have a look at the [other possible endpoints](https://rapidpro.app.medicmobile.org/api/v2/) we might want to pull in and include (or make it easy to include) others if necessary.  For example... I would have though we'd include `definitions` and/or some of the `flow` related endpoints. @marialma and @kitsao would probably be best positioned to know that.

**Comment by mrjones-plip** [comment]
@marialma and @kitsao - can I push on you a bit to define the RapidPro API endpoints we should include?  Maybe we could have a confirmed list beyond just `Contacts`, `Messages` and `Runs` say in a week's time, so done by Apr 13th? While no real rush, be good to knock out this list so this ticket isn't blocked by requirements.

If we need a real time meeting to review, lemme know, otherwise I'll assume y'all are good to go with the info in this, and related, tickets.  

Thanks thanks!

**Comment by marialma** [comment]
Random ask: how is this dealing with RP archiving flows older than a month?

Otherwise: exporting all the definitions would be helpful as well for the Rapidpro Analysis Facilitation Tool. It'll end up being just a ton of JSON but it's definitely a core need for RAFT

**Comment by mrjones-plip** [comment]
After [some good discussion on Slack](https://medic.slack.com/archives/C024KSXQX/p1617889553016700) (private link), here's the feedback on my question to Maria:

* Maria has an awesome [RAFT](https://github.com/medic/medic-impact/tree/327-mm-RAFT/learning_roadmap/patient_messaging/327-RAFT) (private repo) project which looks to be a nice compliment to rapidpro2pg effort especially if it can be ported to use postgres as a back end instead of the interstitial spreadsheets.
* [RapidPro's API](https://rapidpro.io/api/v2/) has 26 odd endpoints, and we think we should incorporate them all into rapidpro2pg (may be some overlap w/ `*_actions` and `*_events`, but TBD)
* There doesn't look be a concern about automatic archiving:
   * "Flows can be archived manually, but it would surprise me if there was anything that did that automatically"
   *  "flow API is filterable on the archived field, I would encourage us to include archived flows"

**Comment by kitsao** [comment]
From several conversations with @mouriceb:
1. we need the full details of a [run](https://textit.in/api/v2/runs). We currently are capturing minimal details. We need all details so that we can tell which runs had a two way interaction with clients. We cannot tell this from the messages.
2. We need to get all flow JSONs to Postgres. Some data such as nodes and values from the runs above can only be referenced from the flow JSON. This will help link responses to runs.

**Comment by mrjones-plip** [comment]
Awesome, thanks for the extra details @kitsao!

**Comment by dianabarsan** [comment]
@kitsao @mrjones-plip 
Can anyone please update about the state of this? Have the changes in [this comment](https://github.com/medic/cht-core/issues/6641#issuecomment-816503022) been implemented? 
Is the code on [rapidpro2pg](https://github.com/medic/rapidpro2pg) ready to be reviewed by a dev?

**Comment by mrjones-plip** [comment]
@dianabarsan  - I don't see any changes to the [rapidpro2pg repo](https://github.com/medic/rapidpro2pg) in the past 6 mo, so I don't think @kitsao's changes in his comments you cited have been made - but I defer to @kitsao !

Further, I don't know if the code in the rapidpro2pg repo has all the features discussed above (specifically including all the endpoints).

Either of these may affect if the code is considered ready to be reviewed by a dev, depending if core eng is expected to add any missing features or if we need that to come from @kitsao .

Let me know if you'd like some PO work on this ticket to make it more mature in any way!

**Comment by dianabarsan** [comment]
I have a question about how we want to introduce flow/run details into PG. 
Would it be preferred to have a "nodes" table to be joined with runs and/or flows to get all info? Or should node data be duplicated in every flow/run ? For runs, the duplication could end up as significant volume, less for flows. 

We could also just store all nodes information directly in the "flow" and join on the "flow" when wanting to get all data for a "run", and not have a "nodes" table at all. 

What is most convenient?

**Comment by mrjones-plip** [comment]
Thanks for the question @dianabarsan -  Deferring to @kitsao  on this.

If no one else know the answer, I'm happy to dig into this to find the answer though!

**Comment by dianabarsan** [comment]
@mrjones-plip we haven't had any updates on this. How should I proceed?

**Comment by mrjones-plip** [comment]
@dianabarsan  - @michaelkohn  and I chatted and he's gonna jump in with suggested next steps since has more background on runs/flows than I do.

**Comment by michaelkohn** [comment]
Just to close the loop on this.

Diana and I chatted, here are some notes/outputs
- One table per RapidPro endpoint
- Each table will have 2 fields, an id field and a jsonb field with all the data 
- The definitions endpoint is a bit of special case in that you have to pass flow id(s) to it. The "productized" rapidpro2pg will do that
- We are _not_ going to pull out node info into different tables since it's easy to do the joins and fine to do it through views
- For this version, we're just going to focus on getting the data into the tables and _not_ dream up a bunch of helper views
- There are a few matviews already done in Kitsao's version ([here](https://github.com/medic/rapidpro2pg/blob/main/app/migrations/001.do.create.sql)) that we can include

**Comment by dianabarsan** [comment]
This is ready for AT on https://github.com/medic/rapidpro2pg/tree/add-more-endpoints
